File: mobile/src/main/java/com/ataulm/chunks/ChunksScreen.java
package com.ataulm.chunks;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;

import java.util.Arrays;
import java.util.List;

import androidx.viewpager.widget.ViewPager;
import butterknife.BindView;
import butterknife.ButterKnife;

public class ChunksScreen extends FrameLayout implements ChunksView {

    @BindView(R.id.chunks_screen_pager_navigation_widget)
    TabsPagerNavigationWidget tabsPagerNavigationWidget;

    @BindView(R.id.chunks_screen_view_pager)
    ViewPager viewPager;

    @BindView(R.id.chunks_screen_empty_input_widget)
    ItemInputWidget itemInputWidget;

    private final OnPageChangeListenerDelegate onPageChangeListenerDelegate = new OnPageChangeListenerDelegate();

    public ChunksScreen(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        View.inflate(getContext(), R.layout.merge_chunks_screen, this);
        ButterKnife.bind(this);
    }

    @Override
    public void display(Chunks chunks, ItemUserInteractions itemUserInteractions, final ItemInputUserInteractions itemInputUserInteractions) {
        if (chunks.input().isPresent()) {
            itemInputWidget.setText(chunks.input().get());
        }

        viewPager.clearOnPageChangeListeners();
        viewPager.addOnPageChangeListener(onPageChangeListenerDelegate);
        viewPager.addOnPageChangeListener(
                new ViewPager.SimpleOnPageChangeListener() {
                    @Override
                    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                        if (positionOffset != 0) {
                            // for now, only want stuff changing when the page has settled
                            return;
                        }

                        Day day = DayToPagePositionMapper.getDayFor(position);
                        itemInputWidget.update(itemInputUserInteractions, day);
                    }
                }
        );

        updateViewPagerWith(chunks, itemUserInteractions);
    }

    private void updateViewPagerWith(Chunks chunks, ItemUserInteractions itemUserInteractions) {
        ChunksPagerAdapter chunksPagerAdapter;
        List<DayUiModel> dayUiModels = createDayUiModelFrom(chunks);
        if (viewPager.getAdapter() == null) {
            chunksPagerAdapter = new ChunksPagerAdapter(itemUserInteractions, viewPager.getResources(), dayUiModels);
            viewPager.setAdapter(chunksPagerAdapter);
            setViewPager(Day.TODAY); // TODO: this happens on rotate, but not necessarily what we want
        } else {
            chunksPagerAdapter = (ChunksPagerAdapter) viewPager.getAdapter();
            chunksPagerAdapter.update(dayUiModels);
        }
        tabsPagerNavigationWidget.bind(viewPager);
    }

    private void setViewPager(Day day) {
        int page = DayToPagePositionMapper.getPageFor(day);
        viewPager.setCurrentItem(page);
    }

    private static List<DayUiModel> createDayUiModelFrom(Chunks chunks) {
        return Arrays.asList(
                new DayUiModel(R.string.days_tabs_today, Day.TODAY, chunks.today()),
                new DayUiModel(R.string.days_tabs_tomorrow, Day.TOMORROW, chunks.tomorrow()),
                new DayUiModel(R.string.days_tabs_sometime, Day.SOMETIME, chunks.sometime())
        );
    }

}

File: mobile/src/main/java/com/ataulm/chunks/ChunksScreen.kt
package com.ataulm.chunks

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import androidx.viewpager.widget.ViewPager
import kotlinx.android.synthetic.main.merge_chunks_screen.view.*
import java.util.*

class ChunksScreen(context: Context, attrs: AttributeSet) : FrameLayout(context, attrs), ChunksView {

    private lateinit var tabsPagerNavigationWidget: TabsPagerNavigationWidget
    private lateinit var viewPager: ViewPager
    private lateinit var itemInputWidget: ItemInputWidget

    private val onPageChangeListenerDelegate = OnPageChangeListenerDelegate()

    override fun onFinishInflate() {
        super.onFinishInflate()
        View.inflate(context, R.layout.merge_chunks_screen, this)
        tabsPagerNavigationWidget = chunks_screen_pager_navigation_widget
        viewPager = chunks_screen_view_pager
        itemInputWidget = chunks_screen_empty_input_widget
    }

    override fun display(chunks: Chunks, itemUserInteractions: ItemUserInteractions, itemInputUserInteractions: ItemInputUserInteractions) {
        chunks.input?.let { itemInputWidget.setText(it) }

        viewPager.clearOnPageChangeListeners()
        viewPager.addOnPageChangeListener(onPageChangeListenerDelegate)
        viewPager.addOnPageChangeListener(
                object : ViewPager.SimpleOnPageChangeListener() {
                    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                        if (positionOffset != 0f) {
                            // for now, only want stuff changing when the page has settled
                            return
                        }

                        val day = DayToPagePositionMapper.getDayFor(position)
                        itemInputWidget.update(itemInputUserInteractions, day)
                    }
                }
        )

        updateViewPagerWith(chunks, itemUserInteractions)
    }

    private fun updateViewPagerWith(chunks: Chunks, itemUserInteractions: ItemUserInteractions) {
        val chunksPagerAdapter: ChunksPagerAdapter
        val dayUiModels = createDayUiModelFrom(chunks)
        if (viewPager.adapter == null) {
            chunksPagerAdapter = ChunksPagerAdapter(itemUserInteractions, viewPager.resources, dayUiModels)
            viewPager.adapter = chunksPagerAdapter
            setViewPager(Day.TODAY) // TODO: this happens on rotate, but not necessarily what we want
        } else {
            chunksPagerAdapter = viewPager.adapter as ChunksPagerAdapter
            chunksPagerAdapter.update(dayUiModels)
        }
        tabsPagerNavigationWidget.bind(viewPager)
    }

    private fun setViewPager(day: Day) {
        val page = DayToPagePositionMapper.getPageFor(day)
        viewPager.currentItem = page
    }

    private fun createDayUiModelFrom(chunks: Chunks): List<DayUiModel> {
        return Arrays.asList(
                DayUiModel(R.string.days_tabs_today, Day.TODAY, chunks.today),
                DayUiModel(R.string.days_tabs_tomorrow, Day.TOMORROW, chunks.tomorrow),
                DayUiModel(R.string.days_tabs_sometime, Day.SOMETIME, chunks.sometime)
        )
    }

}
